doc(proposal): `expectFailure` label and/or matcher (#10)
nodejs/test-runner: proposals/expect-failure-enhancements.md

# Feature proposal: `expectFailure` enhancements

## Summary
Update the `expectFailure` option in `test()` to accept different types of values, enabling both **custom failure messages** and **error validation**. This proposal integrates the requirements from [nodejs/node#61570](https://github.com/nodejs/node/issues/61570), ensuring consistency with `skip`/`todo` while adding robust validation capabilities.

## API & Behavior

The behavior of `expectFailure` is strictly determined by the type of value provided:

> ### 1. String: Failure Reason

Reviewer: ```suggestion
### String: Failure label
```

> When a **non-empty string** is provided, it acts as a documentation message (reason), identical to `skip` and `todo` options.

Reviewer: ```suggestion
When a **non-empty** `string` is provided, it acts as an output label (reason) for that test-case (identical to `skip` and `todo` options).
```

```js
test('fails with a specific reason', {
  expectFailure: 'Bug #123: Feature not implemented yet'
}, () => {
  throw new Error('boom');
});
```
- **Behavior**: The test is expected to fail. The string is treated as a label/reason.

> - **Validation**: None. It accepts *any* error.

Reviewer: nit: disambiguate with bold

```suggestion
- **Validation**: None. It accepts _any_ error.
```

> - **Output**: The reporter displays the string (e.g., `# EXPECTED FAILURE Bug #123...`).

Reviewer: ```suggestion
- **Output**: The reporter displays the string literally (e.g., `# EXPECTED FAILURE Bug #123…`).
```

> ### 2. Matcher: RegExp, Class, or Error Object

Reviewer: nit for the sequence (alphabetical so there's no perceived hierarchy)

```suggestion
### Matcher: Error class, Error-like object, RegExp, or validation function

```

> When a **RegExp**, **Class** (Function), or **Error Object** is provided directly, it acts as the validation logic. This leverages `assert.throws` behavior directly.

Reviewer: ```suggestion
When a `Error` constructor, `Error`-like object, `RegExp`, or validation function is provided, it is treated as validation to match against, mirroring the behaviour of [`assert.throws`](https://nodejs.org/docs/latest-v25.x/api/assert.html#assertthrowsfn-error-message) (possibly leveraging it under the hood).
```

```js
test('fails with matching error (RegExp)', {
  expectFailure: /expected error message/
}, () => {
  throw new Error('this is the expected error message');
});

test('fails with matching error (Class)', {
  expectFailure: RangeError
}, () => {
  throw new RangeError('Index out of bounds');
});
```

> ### 3. Configuration Object: Reason & Validation

Reviewer: ```suggestion
### Configuration Object: Reason & Validation

```

```js
test('fails with reason and specific error', {
  expectFailure: {
    message: 'Bug #123: Edge case behavior', // Reason
    with: /Index out of bounds/              // Validation
  }
}, () => {
  throw new RangeError('Index out of bounds');
});
```
- **Properties**:
    - `message` (String): The failure reason/label (displayed in reporter).
    - `with` (RegExp | Object | Function | Class): Validation logic. This is passed directly to `assert.throws` validation argument, supporting all its capabilities.
- **Behavior**: The test passes **only if** the error matches the `with` criteria.
- **Output**: The reporter displays the `message`.

### Equivalence
The following configurations are equivalent in behavior:

**1. Reason only:**
```js
expectFailure: 'reason'
expectFailure: { message: 'reason' }
```

**2. Validation only:**
```js
expectFailure: /error/
expectFailure: { with: /error/ }
```

**3. Catch-all (Any Error):**
```js
expectFailure: true
```

## Ambiguity Resolution
Potential ambiguity between a **Matcher Object** and a **Configuration Object** is resolved as follows:

> 1.  **String** → Reason.

Reviewer: ```suggestion
1.  **String** → Label.
```

2.  **RegExp** or **Function** → Matcher (Validation).
3.  **Object**:
    *   **Empty Object** (`{}`) → **Error**: throws `ERR_INVALID_ARG_VALUE`.
        ```js
        // Uses Node.js standard error code
        throw new ERR_INVALID_ARG_VALUE(
          'expectFailure',
          expectFailure,
          'must not be an empty object'
        );
        ```

>     *   If the object contains `with` or `message` properties → **Configuration Object**.

Reviewer: I think if it contains either or both but no more

    *   Otherwise → **Matcher Object** (passed to `assert.throws` for property matching).

## Activation & Truthiness
To maintain strict consistency with `todo` and `skip` options:

> *   The feature is **disabled** only if `expectFailure` is `undefined` or `false`.

Reviewer: This isn't mentioned in the API docs, so I'd have to check the implementation, but I would guess this is not correct—any falsy value is "off" (e.g. `skip: null` also disables it).

Reviewer 2: See https://github.com/nodejs/test-runner/pull/10#issuecomment-3827874789

```js
test('fails with matching error (RegExp)', {
  expectFailure: /expected error message/
}, () => {
  throw new Error('this is the expected error message');
});

test('fails with matching error (Class)', {
  expectFailure: RangeError
}, () => {
  throw new RangeError('Index out of bounds');
});
```

### Configuration Object: Reason & Validation

When a **Plain Object** with specific properties (`match`, `label`) is provided, it allows specifying both a failure reason and validation logic simultaneously.

```js
test('fails with reason and specific error', {
  expectFailure: {
    label: 'Bug #123: Edge case behavior', // Reason
    match: /Index out of bounds/              // Validation
  }
}, () => {
  throw new RangeError('Index out of bounds');
});
```
- **Properties**:
    - `label` (String): The failure reason/label (displayed in reporter).
    - `match` (RegExp | Object | Function | Class): Validation logic. This is passed directly to `assert.throws` validation argument, supporting all its capabilities.
- **Behavior**: The test passes **only if** the error matches the `match` criteria.

> - **Output**: The reporter displays the `label`.

Reviewer 3: I think that either one of these properties must be present, or it should throw - iow, using the object form with just a label is fine, because it reduces diff churn if `match` is added or removed later; and using the object form with just `match` is fine, because it reduces diff churn if `label` is added later - but if you have neither, it's far more likely to be a bug than intentional behavior.

(update: i see further down that this is the case, but imo it should also mention here that one of these two properties must be present)

When an `Error` constructor, `Error`-like object, `RegExp`, or validation function is provided, it is treated as validation to match against, mirroring the behaviour of [`assert.throws`](https://nodejs.org/docs/latest-v25.x/api/assert.html#assertthrowsfn-error-message) (possibly leveraging it under the hood).

```js
test('fails with matching error (RegExp)', {
  expectFailure: /expected error message/
}, () => {
  throw new Error('this is the expected error message');
});

test('fails with matching error (Class)', {
  expectFailure: RangeError
}, () => {
  throw new RangeError('Index out of bounds');
});
```

### Configuration Object: Reason & Validation

When a **Plain Object** with specific properties (`match`, `label`) is provided, it allows specifying both a failure reason and validation logic simultaneously.

```js
test('fails with reason and specific error', {
  expectFailure: {
    label: 'Bug #123: Edge case behavior', // Reason
    match: /Index out of bounds/              // Validation
  }
}, () => {
  throw new RangeError('Index out of bounds');
});
```
- **Properties**:
    - `label` (String): The failure reason/label (displayed in reporter).

>     - `match` (RegExp | Object | Function | Class): Validation logic. This is passed directly to `assert.throws` validation argument, supporting all its capabilities.

Reviewer 2: @JakobJingleheimer et. al., when writing [my suggested changes to the PR](https://github.com/nodejs/node/pull/61563#pullrequestreview-3749319585), I noticed `assert.throws` names this field `error`, not `match`. Might it be better to be consistent?

Reviewer: Ah, sure, yes :) `error` was one of the ideas I suggested (but without knowing it was what `assert.throws` uses).

### Flat Options (`expectFailureError`)
It was proposed to split the options into `expectFailure` (reason) and `expectFailureError` (validation).
This was rejected in favor of the nested/polymorphic structure using `with` and `message` properties. This syntax was selected as the preferred choice for its readability and clarity:
*   `with`: Clearly indicates "fails **with** this error" (Validation).
*   `message`: Clearly indicates the **reason** or label for the expected failure.
This approach keeps related configuration grouped without polluting the top-level options namespace.

## Implementation Details

### Validation Logic
The implementation leverages `assert.throws` internally to perform error validation.
- If `expectFailure` is a Matcher (RegExp, Class, Object), it is passed as the second argument to `assert.throws(fn, expectFailure)`.
- If `expectFailure` is a Configuration Object, `expectFailure.with` is passed to `assert.throws`.

## Edge Cases

### Empty String (`expectFailure: ''`)
To maintain consistency with `todo` and `skip` options, an empty string is treated as **truthy** (enabled).
*   `expectFailure: ''` behaves like `expectFailure: true`.

> *   The feature is **enabled** (expects failure), but without a specific reason label.

Reviewer 2: I think the implementation should replicate the todo/skip logic exactly: *any* value other than `undefined` or `false` is treated as true, unless it is decided to change todo/skip as well.

});
```
- **Behavior**: The test is expected to fail. The string is treated as a label/reason.
- **Validation**: None. It accepts _any_ error.
- **Output**: The reporter displays the string literally (e.g., `# EXPECTED FAILURE Bug #123…`).

### Matcher: Error constructor, Error-like object, RegExp, or validation function

> When a `Error` constructor, `Error`-like object, `RegExp`, or validation function is provided, it is treated as validation to match against, mirroring the behaviour of [`assert.throws`](https://nodejs.org/docs/latest-v25.x/api/assert.html#assertthrowsfn-error-message) (possibly leveraging it under the hood).

Reviewer: nit 😅

```suggestion
When an `Error` constructor, `Error`-like object, `RegExp`, or validation function is provided, it is treated as validation to match against, mirroring the behaviour of [`assert.throws`](https://nodejs.org/docs/latest-v25.x/api/assert.html#assertthrowsfn-error-message) (possibly leveraging it under the hood).
```

          'must not be an empty object'
        );
        ```
    *   If the object contains **only** `match` and/or `label` properties → **Configuration Object**.
    *   Otherwise → **Matcher Object** (passed to `assert.throws` for property matching).

## Activation & Truthiness
To maintain strict consistency with `todo` and `skip` options:
* The feature is **disabled** if `expectFailure` is **falsy** (e.g., `false`, `null`, `undefined`, `0`, `''`).

> * **Truthy values** enable the feature.

Reviewer: Whoops, sorry, I was wrong about this: https://github.com/nodejs/test-runner/pull/10#discussion_r2756325993

You can revert this part of the change—it was good/correct before.
